### ISSUE TITLE
Allowing users to select NDT server will be top

### ISSUE THREAD

**Issue by bassosimone** [ooni/probe#820]
I was browsing through the OONI Probe reviews on the Play Store, and noticed a user who expressed this specific feature request. I'm not sure this can be done easily and soon, but let's track it.

**Comment by elioqoshi** [comment]
I suppose it would be sensible to have this in 2.1 and have a dropdown menu which fetches the server list from M-Lab?

**Comment by lorenzoPrimi** [comment]
Last time we spoke about this I moved into the "Never" column on Trello as was not something we need. Do you recall the reason @bassosimone @hellais ?

**Comment by elioqoshi** [comment]
Would be good to get that context so we can close this or know what is needed going forward.

**Comment by lorenzoPrimi** [comment]
@bassosimone is it a never or a later?

**Comment by bassosimone** [comment]
This is complicated by the way in which mlab-ns works. It's theoretically possible to enumerate all servers, but placing them on the map is tricky. I suggest to keep it open with low priority.

Also, looping in @stephen-soltesz, because I know he's taking care of mlab-ns and he may be interested to know what feature request we did indirectly get for it. (And, maybe, there is also a smart mlab-ns way of solving this, and I'm just not aware of that.)

**Comment by stephen-soltesz** [comment]
Do you mean like this? This returns all sites including some location metadata (country & city) http://mlab-ns.appspot.com/ndt_ssl?policy=all

**Comment by bassosimone** [comment]
@stephen-soltesz yeah, that is probably good. From the city, we can theoretically find out a way to put a marker on the map and let the user choose the marker. Thanks!

**Comment by bassosimone** [comment]
So, rather than using a map, we can use `curl 'https://locate.measurementlab.net/ndt7?policy=geo_options'` to just get a few options and we can ask the user how many of them to enable.

**Comment by darkk** [comment]
> `https://locate.measurementlab.net/ndt7?policy=geo_options`

FTR, it currently offers me four options:
1. Stockholm (arn06)
2. Stockholm (arn02)
3. Stockholm (arn03)
4. Stockholm (arn04)

So it may be not that diverse set in some (maybe, edge) cases.

**Comment by bassosimone** [comment]
@darkk the `geo_options` policy gives you some of the geographically closest to you servers, with the uncertainty caused by the fact that your location estimation may be inaccurate. In light of that, it makes sense that you are given a bunch of locations in Stockholm.

What should probably be stressed is that each of those sites is a different ISP. Therefore, I find it reasonably likely that you reach them over different paths. This is potentially interesting for research and potentially painful if you reason in terms of consumer facing speed test. There have been several instances where interconnection issues where spotted by testing all available servers. OTOH, since the default server selection selects on server at random among the ones that are listed inside of the `geo_options` list (roughly), it may sometimes be that you choose the testing server that have the least performance from your vantage point.

The criteria for servers placement AFAIK is to place servers in locations where there is content, such that accessing the servers reflects the typical experience of using the public internet, for a definition of public internet where we exclude the cloud and private CDNs for the time being.

I don't know what was the reason why the original poster on the Play Store wanted to see a list of servers, because it wasn't possible to gather further details. In my view that is important though because in a bunch of cases I did find myself unhappy with the defaults returned by the locate service and would have fancied manually choosing which server. I guess what complicated the issue is the quite understandable tension between having the feature and making it user friendly.

**Comment by bassosimone** [comment]
It has ben suggested that a nice way to avoid overwhelming users with information would be to have a mode where we test all the available servers rather than the closest one. I like this idea because it does not pass the burden of choosing which server _on all users_. OTOH, there is this issue that, when something is not the default, many people will not be using it.

**Comment by torch016** [comment]
> It has ben suggested that a nice way to avoid overwhelming users with information would be to have a mode where we test all the available servers rather than the closest one. I like this idea because it does not pass the burden of choosing which server _on all users_. OTOH, there is this issue that, when something is not the default, many people will not be using it.

Shouldn't server close to the user be of more priority than those farther away?
